### PR TITLE
Pivot wider data table

### LINKED ISSUE #1066: pivot_wider does not work well with data.table
Per this [StackOverflow Question](https://stackoverflow.com/questions/65047707/why-doesnt-pivot-wider-work-on-a-data-table), it appears that there can be errors when `data.table` is loaded and `data.table` methods are dispatched on `data.table` objects (i.e., `[.data.frame` vs. `[.data.table`.

``` r
library(data.table)
library(tidyr)
library(dplyr)

df <- data.frame(names=letters[1:3],values=1:3)
df %>% pivot_wider(names_from=names,values_from=values)
#> # A tibble: 1 x 3
#>       a     b     c
#>   <int> <int> <int>
#> 1     1     2     3

dt <- data.table(names1=letters[1:3],values1=1:3)
pivot_wider(dt, names_from=names1,values_from=values1)
#> Error in data.frame(row = row_id, col = col_id): arguments imply differing number of rows: 0, 3
```
This is the traceback:
>5: stop(gettextf("arguments imply differing number of rows: %s", 
       paste(unique(nrows), collapse = ", ")), domain = NA)
4: data.frame(row = row_id, col = col_id)
3: pivot_wider_spec(data, spec, !!id_cols, names_repair = names_repair, 
       values_fill = values_fill, values_fn = values_fn)
2: pivot_wider.data.frame(dt, names_from = names1, values_from = values1)
1: pivot_wider(dt, names_from = names1, values_from = values1)

I believe this is the problem:

https://github.com/tidyverse/tidyr/blob/501a757bf983a3eb6e82f53173672a92d828236f/R/pivot-wide.R#L224

To make it agnostic between ```data.table``` and ```data.frame``` methods, this should work:

```r
df_rows <- data[, key_vars, drop = FALSE]
```

As is, ```dt[cols]``` would be equivalent to ```df[cols, ]```. I can make a PR if you'd like.

### PR DESCRIPTION
**Pull request by mgirlich**
Fixes #1066

### CODE REVIEW

File: tests/testthat/test-pivot-wide.R [hunk 1]
```diff
@@ -65,6 +65,16 @@ test_that("data frame columns pivot correctly", {
   expect_equal(out$b$y, 7:8)
 })
 
+test_that("works with data.table and empty key_vars", {
+  skip_if_not_installed("data.table")
```
Comment (hadley) [inline tests/testthat/test-pivot-wide.R:69]: We assume that suggested packages are installed for development, so you can remove this.

**Comment by ColeMiller1** [comment]
@mgirlich . I spent some time researching the issue and solution. I feel like this is my PR to make.

**Comment by mgirlich** [comment]
@ColeMiller1 sorry, didn't mean to take this away from you. Please go ahead and make a PR.

**Comment by hadley** [comment]
@ColeMiller1 do you still want to make a PR for this?

**Comment by ColeMiller1** [comment]
Sorry for being a bit of a jerk. Please move forward with the PR and thanks
for the fix!

**Comment by hadley** [comment]
Worth noting that dtplyr now includes a translation from `pivot_wider()` to `dcast()`.

**Pull request merged**